### ISSUE TITLE
Request: option to not pull images if already present

### ISSUE THREAD

**Issue by bgehman** [rancher/rke#137]
**RKE version:**
`rke version v0.0.8-dev`

**Docker version: (`docker version`,`docker info` preferred)**
```
$ docker info

Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 12
Server Version: 1.12.6
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 67
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host null bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-103-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.855 GiB
Name: ubuntu
ID: W4UI:5PVV:WIJ3:BUKF:SOKU:2F7F:UM7A:PXS6:GUOM:YCJP:2AQ6:L6VR
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
```
$ cat /etc/os-release

NAME="Ubuntu"
VERSION="16.04.3 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.3 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

```
$ uname -r
4.4.0-103-generic
```

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
KVM VM on laptop -- simulating a network isolated "on-prem" installation.

**cluster.yml file:**
```
nodes:
- address: 192.168.123.27 
  role:
  - controlplane
  - worker
  - etcd
  user: ubuntu 
  docker_socket: /var/run/docker.sock

services:

  etcd:
    image: quay.io/coreos/etcd:v3.2

  kube-api:
    image: rancher/k8s:v1.8.3-rancher2
    service_cluster_ip_range: 10.233.0.0/18

  kube-controller:
    image: rancher/k8s:v1.8.3-rancher2
    cluster_cidr: 10.233.64.0/18
    service_cluster_ip_range: 10.233.0.0/18

  scheduler:
    image: rancher/k8s:v1.8.3-rancher2

  kubelet:
    image: rancher/k8s:v1.8.3-rancher2
    cluster_domain: cluster.local
    infra_container_image: gcr.io/google_containers/pause-amd64:3.0
    cluster_dns_server: 10.233.0.3

  kubeproxy:
    image: rancher/k8s:v1.8.3-rancher2

network:
  plugin: flannel
  options:
    flannel_image: quay.io/coreos/flannel:v0.9.1
    flannel_cni_image: quay.io/coreos/flannel-cni:v0.2.0

auth:
  strategy: x509

system_images:
  alpine: alpine:3.7
  nginx_proxy: rancher/rke-nginx-proxy:0.1.0
  cert_downloader: rancher/rke-cert-deployer:0.1.0
  kubedns_image: gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.5
  dnsmasq_image: gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64:1.14.5
  kubedns_sidecar_image: gcr.io/google_containers/k8s-dns-sidecar-amd64:1.14.5
  kubedns_autoscaler_image: gcr.io/google_containers/cluster-proportional-autoscaler-amd64:1.0.0

ssh_key_path: ~/.ssh/rke_rsa
```

**Steps to Reproduce:**
Trying to see if we can use rke to deploy kubernetes to network isolated systems.  These are typical "on-prem" systems that have no connectivity to the outside world (customer security requirements).

On a Network Isolated VM, manually preload all required docker images from tar/gz (ie, using `docker load`).  Current state of the Ubuntu test system shows all the needed docker images have been preloaded:

```
$ docker images

REPOSITORY                                                       TAG                 IMAGE ID            CREATED             SIZE
quay.io/coreos/etcd                                              v3.2                83f6f4862815        7 days ago          37.18 MB
alpine                                                           3.7                 e21c333399e0        11 days ago         4.144 MB
rancher/rke-nginx-proxy                                          0.1.0               63ad36f1e7c2        13 days ago         39.13 MB
rancher/rke-cert-deployer                                        0.1.0               c8907e804cfe        13 days ago         9.088 MB
rancher/k8s                                                      v1.8.3-rancher2     bbbe40353d71        3 weeks ago         1.543 GB
quay.io/coreos/flannel                                           v0.9.1              2b736d06ca4c        3 weeks ago         51.34 MB
gcr.io/google_containers/k8s-dns-sidecar-amd64                   1.14.5              fed89e8b4248        10 weeks ago        41.82 MB
gcr.io/google_containers/k8s-dns-kube-dns-amd64                  1.14.5              512cd7425a73        10 weeks ago        49.39 MB
gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64             1.14.5              459944ce8cc4        10 weeks ago        41.42 MB
quay.io/coreos/flannel-cni                                       v0.2.0              7252edf978c0        3 months ago        49.79 MB
gcr.io/google_containers/cluster-proportional-autoscaler-amd64   1.0.0               e183460c484d        13 months ago       48.16 MB
gcr.io/google_containers/pause-amd64                             3.0                 99e59f495ffa        19 months ago       746.9 kB
```

**Results:**
Using rke to bring up the cluster, fails pretty quickly trying to pull docker images that are already loaded.  Since the VM is network isolated, the `docker pull` (equivalent) will always fail.  Request to add an option similar to Kubernetes' `IfNotPresent` ImagePullPolicies (to not attempt to pull an image if it is already loaded), or gladly accept suggestions on other ways to install kubernetes on network-isolated environments.

```
$ ./rke_0.0.8 up

INFO[0000] Building Kubernetes cluster                  
INFO[0000] [ssh] Setup tunnel for host [192.168.123.27] 
INFO[0000] [ssh] Setup tunnel for host [192.168.123.27] 
INFO[0000] [ssh] Setup tunnel for host [192.168.123.27] 
INFO[0000] [certificates] Generating kubernetes certificates 
INFO[0000] [certificates] Generating CA kubernetes certificates 
INFO[0000] [certificates] Generating Kubernetes API server certificates 
INFO[0001] [certificates] Generating Kube Controller certificates 
INFO[0001] [certificates] Generating Kube Scheduler certificates 
INFO[0002] [certificates] Generating Kube Proxy certificates 
INFO[0002] [certificates] Generating Node certificate   
INFO[0002] [certificates] Generating admin certificates and kubeconfig 
INFO[0002] [reconcile] Reconciling cluster state        
INFO[0002] [reconcile] This is newly generated cluster  
INFO[0002] [certificates] Deploying kubernetes certificates to Cluster nodes 
INFO[0006] Successfully Deployed local admin kubeconfig at [./.kube_config_cluster.yml] 
INFO[0006] [certificates] Successfully deployed kubernetes certificates to Cluster nodes 
INFO[0006] [etcd] Building up Etcd Plane..              
INFO[0006] [etcd] Pulling Image on host [192.168.123.27] 
FATA[0006] [etcd] Failed to bring up Etcd Plane: Can't pull Docker image [quay.io/coreos/etcd:v3.2] for host [192.168.123.27]: Error response from daemon: Get https://quay.io/v1/_ping: dial tcp: lookup quay.io on 192.168.123.1:53: server misbehaving 
```

**Comment by galal-hussein** [comment]
@bgehman This is an interesting issue, we enabled Airgap installations by adding `system_images:`, but we assumed that user will set up a private registry that contain the images and accessible from all hosts, instead of loading all the images on every host, i think this workaround will work successfully with your situation.

**Comment by bgehman** [comment]
Thanks for the suggestion @galal-hussein , will look into it.

**Comment by ibuildthecloud** [comment]
@galal-hussein This really should be the default behavior to pull only if not found.  Don't pull always.  There are plenty of use cases where assuming a valid registry exists is just not feasible.

**Comment by bgehman** [comment]
@galal-hussein @ibuildthecloud It does get messy if you are using `latest` tagged images, or other upstream images where existing tags are constantly moved to point to newer versions.  Perhaps you do want to attempt a pull in those cases...  I'd be content with just trying to pull, and if it fails but the image already exist locally: go with what you got (maybe log a warning instead).  It should  really be a failure unless the container can't start...

**Comment by superseb** [comment]
@bgehman Current logic in the PR is checking for local image first, if found, it will be used, if not, we pull. The tag `latest` or moved existing tags should be avoided in general, and in your case, you are manually loading the images to the machines which makes you in control of the state of the images on the machines.

Using a pull policy parameter was considered, but we like to keep options to a minimal to not confuse users. If you have an example case of why we should reconsider, please let us know. I think the logic in the PR will help you with your use-case.

**Comment by bgehman** [comment]
@superseb the proposed changes will help my use-case.  Many thanks to all who are working on it.  👍   The default stance of rke README has etcd using `quay.io/coreos/etcd:latest`.  So original behavior was that etcd would always be upgraded (if/when new `latest` image is cut), new behavior is that rke will continue to use whatever version of etcd was downloaded when the cluster was originally built.  I agree that it is not wise to use `latest` for anything other than development environments.

**Comment by bgehman** [comment]
Verified fixed in rke v0.0.9:

```
...
INFO[0002] [certificates] Checking image [rancher/rke-cert-deployer:0.1.0] on host [192.168.123.160] 
INFO[0002] [certificates] No pull necessary, image [rancher/rke-cert-deployer:0.1.0] exists on host [192.168.123.160]
...
```

Thanks!
